state/address.go

```diff
@@ -88,59 +88,140 @@ func (st *State) Addresses() ([]string, error) {
 	return nil
 }
 
-// APIHostPorts returns the API addresses as set by SetAPIHostPorts.
+// getOpsForHostPortsChange returns a slice of operations used to update an
+// API host/collection in the DB.
+// If the current document indicates the same host/port collection as the
+// input, no operations are returned.
+func (st *State) getOpsForHostPortsChange(
+	mc mongo.Collection, key string, newHostPorts [][]network.HostPort) ([]txn.Op, error) {
+	var ops []txn.Op
+
+	// Retrieve the current document.
+	var extantHostPortDoc apiHostPortsDoc
+	err := mc.Find(bson.D{{"_id", key}}).One(&extantHostPortDoc)
+	if err != nil {
+		return ops, err
+	}
+
+	// Queue an update operation if the host/port collections differ.
+	extantHostPorts := networkHostsPorts(extantHostPortDoc.APIHostPorts)
+	if !hostsPortsEqual(newHostPorts, extantHostPorts) {
+		ops = []txn.Op{{
+			C:  controllersC,
+			Id: key,
+			Assert: bson.D{{
+				"txn-revno", extantHostPortDoc.TxnRevno,
+			}},
+			Update: bson.D{{
+				"$set", bson.D{{"apihostports", fromNetworkHostsPorts(newHostPorts)}},
+			}},
+		}}
+		logger.Debugf("setting %s: %v", key, newHostPorts)
+	}
+	return ops, nil
+}
+
+// filterHostPortsForManagementSpace filters the collection of API addresses
+// based on the configured management space for the controller.
+// If there is no space configured, or if the one of the slices is filtered down
+// to zero elements, just use the unfiltered slice for safety - we do not
+// want to cut off communication to the controller based on erroneous config.
+func (st *State) filterHostPortsForManagementSpace(apiHostPorts [][]network.HostPort) ([][]network.HostPort, error) {
+	config, err := st.ControllerConfig()
+	if err != nil {
+		return nil, err
+	}
+
+	var hostPortsForAgents [][]network.HostPort
+	if mgmtSpace := config.JujuManagementSpace(); mgmtSpace != "" {
+		hostPortsForAgents = make([][]network.HostPort, len(apiHostPorts))
+		sp := network.SpaceName(mgmtSpace)
+		for i := range apiHostPorts {
+			if filtered, ok := network.SelectHostsPortBySpaces(apiHostPorts[i], sp); ok {
+				hostPortsForAgents[i] = filtered
+			} else {
+				hostPortsForAgents[i] = apiHostPorts[i]
+			}
+		}
+	} else {
+		hostPortsForAgents = apiHostPorts
+	}
+	return hostPortsForAgents, nil
+}
+
+// APIHostPorts returns the collection of *all* API addresses as set by SetAPIHostPorts.
 func (st *State) APIHostPorts() ([][]network.HostPort, error) {
+	return st.apiHostPortsForKey(apiHostPortsKey)
+}
+
+// APIHostPortsForAgents returns the collection of API addresses that should be used
+// by agents.
+// If there is no management network space configured for the controller
+// or if the space is misconfigured, the return will be the same as APIHostPorts.
+// Otherwise the returned addresses will correspond with the management net space.
+func (st *State) APIHostPortsForAgents() ([][]network.HostPort, error) {
+	return st.apiHostPortsForKey(apiHostPortsForAgentsKey)
```

Comment:
Shouldn't we have a direct fallback to APIHostPorts here if apiHostPortsForAgentsKey is empty? What will happen during upgrade?

Reply:
Good point. I will accommodate this.